File: src/utils/landmarks.ts
import { LANDMARK_IDXS_LEFT_DOM, LANDMARK_IDXS_RIGHT_DOM, LEFT_HAND_IDXS, RIGHT_HAND_IDXS } from "./landmarkConstants";
import { INPUT_SIZE } from "../../src/config"; // add INPUT_SIZE in a config file

/**
 * Preprocess a sequence of 543x3 landmarks for the ML model
 * - seq: array of frames, each frame is 543x3
 * - returns:
 *    frames: [1, INPUT_SIZE, N]  (batch dimension included)
 *    frameIdxs: [1, INPUT_SIZE]
 */
export function preprocessLiveSequence(seq: number[][][]): { frames: number[][][][]; frameIdxs: number[][] } | null {
    if (!seq.length) return null;

    const data = seq.map(f => f.map(v => v.slice())); // deep copy

    // Determine dominant hand
    const leftSum = data.map(f => LEFT_HAND_IDXS.reduce((sum, idx) => sum + f[idx].filter(v => !isNaN(v)).length, 0));
    const rightSum = data.map(f => RIGHT_HAND_IDXS.reduce((sum, idx) => sum + f[idx].filter(v => !isNaN(v)).length, 0));
    const leftDom = leftSum.reduce((a, b) => a + b, 0) >= rightSum.reduce((a, b) => a + b, 0);

    const idxs = leftDom ? LANDMARK_IDXS_LEFT_DOM : LANDMARK_IDXS_RIGHT_DOM;
    const handIdxs = leftDom ? LEFT_HAND_IDXS : RIGHT_HAND_IDXS;

    // Filter frames with at least one hand landmark present
    const validFrames: number[] = [];
    data.forEach((f, i) => {
        const handNonNaN = handIdxs.reduce((sum, idx) => sum + f[idx].filter(v => !isNaN(v)).length, 0);
        if (handNonNaN > 0) validFrames.push(i);
    });

    if (!validFrames.length) return null;

    let filteredData = validFrames.map(i => idxs.map(idx => data[i][idx])); // [numValidFrames, N, 3]

    // Create frameIdxs
    let frameIdxs = validFrames.map((i, _, arr) => i - arr[0]);

    // ----- PAD OR SLICE TO INPUT_SIZE -----
    if (filteredData.length < INPUT_SIZE) {
        const padCount = INPUT_SIZE - filteredData.length;
        const padFrame = filteredData[filteredData.length - 1].map(f => f.map(() => 0)); // zero pad
        for (let i = 0; i < padCount; i++) filteredData.push(padFrame);
        frameIdxs = [...frameIdxs, ...Array(padCount).fill(-1)];
    } else if (filteredData.length > INPUT_SIZE) {
        const sel = Array.from({ length: INPUT_SIZE }, (_, i) =>
        Math.floor((i / (INPUT_SIZE - 1)) * (filteredData.length - 1))
        );
        filteredData = sel.map(i => filteredData[i]);
        frameIdxs = sel.map(i => frameIdxs[i]);
    }

    // Replace NaNs with 0
    filteredData = filteredData.map(frame => frame.map(lm => lm.map(v => isNaN(v) ? 0 : v)));

    return {
        frames: [filteredData],  // shape [1, INPUT_SIZE, 543, 3]
        frameIdxs: [frameIdxs],  // shape [1, INPUT_SIZE]
    };


}

// Converts MediaPipe results into a 543x3 array
export function extract543Landmarks(faceRes: any, handRes: any, poseRes: any): number[][] {
  const landmarks: number[][] = Array(543)
    .fill(0)
    .map(() => [NaN, NaN, NaN]);

  // Face
  if (faceRes.multiFaceLandmarks?.length) {
    for (let i = 0; i < faceRes.multiFaceLandmarks[0].length; i++) {
      const lm = faceRes.multiFaceLandmarks[0][i];
      landmarks[i] = [lm.x, lm.y, lm.z];
    }
  }

  // Pose (first 10 landmarks)
  if (poseRes.poseLandmarks) {
    for (let i = 0; i < 10; i++) {
      const lm = poseRes.poseLandmarks[i];
      landmarks[502 + i] = [lm.x, lm.y, lm.z];
    }
  }

  // Hands
  if (handRes.multiHandLandmarks?.length) {
    for (let h = 0; h < handRes.multiHandLandmarks.length; h++) {
      const handLms = handRes.multiHandLandmarks[h];
      const handedness = handRes.multiHandedness[h].label; // "Left" or "Right"
      const baseIdx = handedness === "Left" ? 468 : 522;

      for (let i = 0; i < handLms.length; i++) {
        const lm = handLms[i];
        landmarks[baseIdx + i] = [lm.x, lm.y, lm.z];
      }
    }
  }

  return landmarks;
}

File: backend/inference/landmarks.py
import numpy as np
# 40 face landmarks
LIPS_IDXS = np.array([
    61, 185, 40, 39, 37, 0, 267, 269, 270, 409,
    291, 146, 91, 181, 84, 17, 314, 405, 321, 375,
    78, 191, 80, 81, 82, 13, 312, 311, 310, 415,
    95, 88, 178, 87, 14, 317, 402, 318, 324, 308
])
# 21 landmarks for left hand
LEFT_HAND_IDXS = np.array([
    468, 469, 470, 471, 472, 473, 474, 475, 476, 477, 
    478, 479, 480, 481, 482, 483, 484, 485, 486, 487, 
    488
])
# 21 landmarks for right hand
RIGHT_HAND_IDXS = np.array([
    522, 523, 524, 525, 526, 527, 528, 529, 530, 531, 
    532, 533, 534, 535, 536, 537, 538, 539, 540, 541, 
    542
])
# 5 pose landmarks
POSE_IDXS = np.array([502, 503, 504, 505, 506]) 

# combine all landmarks --> 40 + 21 + 21 + 5 = 87 landmarks
LANDMARK_IDXS = np.concatenate([LIPS_IDXS, LEFT_HAND_IDXS, RIGHT_HAND_IDXS, POSE_IDXS]) 

def process_landmarks(rgb, face_mesh, hands, pose):
    return (
        face_mesh.process(rgb),
        hands.process(rgb),
        pose.process(rgb),
    )

# extracting landmarks
def extract_87_landmarks(face_res, hand_res, pose_res):
    # initialize a placeholder for 87 landmarks (x, y, z for each)
    landmarks = np.full((87, 3), np.nan, dtype=np.float32)

    if face_res.multi_face_landmarks: # face
         # only take first 40 landmarks for face
        for i, lm in enumerate(face_res.multi_face_landmarks[0].landmark[:40]):
            landmarks[i] = [lm.x, lm.y, lm.z]

    if pose_res.pose_landmarks: # pose
        # only take first 5 landmarks for pose
        for i, lm in enumerate(pose_res.pose_landmarks.landmark[:5]):
            landmarks[40 + i] = [lm.x, lm.y, lm.z]

    if hand_res.multi_hand_landmarks: # hands
        for hand_lms, handedness in zip(hand_res.multi_hand_landmarks, hand_res.multi_handedness):
            if handedness.classification[0].label == "Left":
                for i, lm in enumerate(hand_lms.landmark):
                    # left hand goes from index 45
                    landmarks[45 + i] = [lm.x, lm.y, lm.z]
            else:  # right hand
                for i, lm in enumerate(hand_lms.landmark):
                    # right hand goes from index 66
                    landmarks[66 + i] = [lm.x, lm.y, lm.z]

    return landmarks
